runtime/major_gc.c

```diff
@@ -241,7 +241,16 @@ Caml_inline void mark_stack_push(struct mark_stack* stk, value block,
   for (i = offset; i < end; i++) {
     v = Field(block, i);
 
-    if (Is_block(v) && !Is_black_val(v))
+    if (Is_block(v) &&
+#ifdef NO_NAKED_POINTERS
+        /* With NO_NAKED_POINTERS, every block has a valid header. */
+        !Is_young(v) &&
+#else
+        /* In the presence of naked pointers, only blocks in the heap are
+         * guaranteed to have a valid header. */
+        Is_in_heap(v) &&
```

Comment:
Again, the comment is off.  Only blocks that satisfy the `Is_in_value_area` test have valid headers, but these include minor heap blocks, major heap blocks, and static data.  Here `Is_in_heap` is tested, not `Is_in_value_area`.